development_bank_wales/analysis/Initial Upgradability Analysis.py

```diff
@@ -0,0 +1,634 @@
+# We measure upgradability based on whether a property has had an improvement in a specific category or whether there has been a recommendation.
+#
+# The most upgrades happen in the category 'lighting' since it's the easiest one. There are hardly any (recorded) upgrades done to the floor of a property, as it is a real hassle.
+#
+# Mainheat, hot water and lighting are most frequently recommended to be upgraded. When only looking at properties with specific recommendation, then lighting, roof and windows are more frequently upgraded.
+
+# %%
+first_wales_w_upgrades = upgrades.get_upgrade_features(
+    first_wales, latest_wales, keep="first"
+)
+
+latest_wales_w_upgrades = upgrades.get_upgrade_features(
+    first_wales, latest_wales, keep="latest", verbose=False
+)
+
+# %%
+cats = ["ROOF", "WINDOWS", "WALLS", "FLOOR", "LIGHTING", "HOT_WATER", "MAINHEAT"]
+
+
+upgrade_df = pd.DataFrame(
+    index=cats,
+    columns=[
+        "Recommendations (%)",
+        "Upgrades (%)",
+        "Recommendations or Upgrades (%)",
+        "Recommendations and Upgrades (%)",
+        "Recommendations and Upgrades (% of all with recs)",
+        "Upgradability (mean)",
+    ],
+)
+upgrade_df = upgrade_df.astype(float)
+
+
+for cat in cats:
+
+    total_props = first_wales_w_upgrades.shape[0]
+    total_props_with_rec = first_wales_w_upgrades.loc[
+        (first_wales_w_upgrades["{}_RECOMMENDATION".format(cat)])
+    ].shape[0]
+    total_props_with_upgr = first_wales_w_upgrades.loc[
+        (first_wales_w_upgrades["{}_EFF_DIFF".format(cat)] > 0)
+    ].shape[0]
+
+    props_w_upgr_and_rec = first_wales_w_upgrades.loc[
+        (
+            (first_wales_w_upgrades["{}_EFF_DIFF".format(cat)] > 0)
+            & (first_wales_w_upgrades["{}_RECOMMENDATION".format(cat)])
+        )
+    ].shape[0]
+
+    props_w_upgr_or_rec = first_wales_w_upgrades.loc[
+        (
+            (first_wales_w_upgrades["{}_EFF_DIFF".format(cat)] > 0)
+            | (first_wales_w_upgrades["{}_RECOMMENDATION".format(cat)])
+        )
+    ].shape[0]
+
+    first_wales_w_upgrades["{}_UPGRADABILITY".format(cat)] = (
+        first_wales_w_upgrades["{}_EFF_DIFF".format(cat)] > 0.0
+    ) | (first_wales_w_upgrades["{}_RECOMMENDATION".format(cat)])
+
+    upgrade_dict = {
+        "Recommendations (%)": total_props_with_rec / total_props * 100,
+        "Upgrades (%)": total_props_with_upgr / total_props * 100,
+        "Recommendations or Upgrades (%)": props_w_upgr_or_rec / total_props * 100,
+        "Recommendations and Upgrades (%)": props_w_upgr_and_rec / total_props * 100,
+        "Recommendations and Upgrades (% of all with recs)": props_w_upgr_and_rec
+        / total_props_with_rec
+        * 100,
+        "Upgradability (mean)": first_wales_w_upgrades[
+            "{}_UPGRADABILITY".format(cat)
+        ].mean()
+        * 100,
+    }
+
+    upgrade_df.loc[cat] = pd.Series(upgrade_dict)
+
+
+upgrade_df = round(upgrade_df)
+upgrade_df.head(10)
```

Comment:
There's only 7 lines (same as the number of categories):
```suggestion
upgrade_df
```